File: insurance-advice/local_registry.py
"""Run the real registry logic locally so the recipe verifies with no network.

Starts a small HTTP server backed by the actual registry.js via Node, and
points the client at it. That way what the tests prove here is what the
deployed Worker does, rather than a Python reimplementation of it.
"""
import json, os, subprocess, tempfile, threading, time
from http.server import BaseHTTPRequestHandler, HTTPServer

BACKEND = "/home/claude/stub/backend"


class _Handler(BaseHTTPRequestHandler):
    registry_proc = None

    def log_message(self, *a):
        pass

    def do_POST(self):
        n = int(self.headers.get("Content-Length", 0))
        body = self.rfile.read(n).decode()
        out = _call_node(self.path, body)
        self._send(out)

    def do_GET(self):
        out = _call_node(self.path, None)
        self._send(out)

    def _send(self, obj):
        data = json.dumps(obj).encode()
        self.send_response(200)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(data)))
        self.end_headers()
        self.wfile.write(data)


_STATE = {"db": None}


def _call_node(path, body):
    """One Node process per call, sharing a sqlite file so state persists."""
    script = f"""
import {{ Registry }} from '{BACKEND}/registry.js';
import {{ makeLocalDb }} from '{BACKEND}/local-db.js';
import fs from 'fs';
const db = makeLocalDb('{_STATE["db"]}');
const wkRaw = fs.existsSync('{_STATE["db"]}.key')
  ? JSON.parse(fs.readFileSync('{_STATE["db"]}.key','utf8')) : null;
let wk;
if (wkRaw) {{
  wk = await crypto.subtle.importKey('jwk', wkRaw, {{name:'Ed25519'}}, true, ['sign']);
}} else {{
  const kp = await crypto.subtle.generateKey({{name:'Ed25519'}}, true, ['sign','verify']);
  const jwk = await crypto.subtle.exportKey('jwk', kp.privateKey);
  fs.writeFileSync('{_STATE["db"]}.key', JSON.stringify(jwk));
  wk = kp.privateKey;
}}
const reg = new Registry(db, wk);
const path = {json.dumps(path)};
const body = {json.dumps(body) if body else 'null'};
let out;
if (path === '/register') out = await reg.registerOperator(JSON.parse(body));
else if (path === '/issue') out = await reg.issue(JSON.parse(body));
else if (path.startsWith('/resolve/') || path.startsWith('/check/')) out = await reg.resolve(path.split('/').pop());
else out = {{ ok: true }};
console.log(JSON.stringify(out));
"""
    with tempfile.NamedTemporaryFile("w", suffix=".mjs", delete=False) as f:
        f.write(script)
        p = f.name
    try:
        r = subprocess.run(["node", p], capture_output=True, text=True, timeout=30)
        line = r.stdout.strip().splitlines()[-1]
        return json.loads(line)
    finally:
        os.unlink(p)


def start(port=8799):
    _STATE["db"] = tempfile.mktemp(suffix=".sqlite")
    srv = HTTPServer(("127.0.0.1", port), _Handler)
    t = threading.Thread(target=srv.serve_forever, daemon=True)
    t.start()
    time.sleep(0.2)
    return srv, f"http://127.0.0.1:{port}"


def stop(srv):
    srv.shutdown()
    for suffix in ("", ".key"):
        try:
            os.unlink(_STATE["db"] + suffix)
        except Exception:
            pass

File: vercel-ai-sdk/local-registry.mjs
// Runs the real Stub registry logic in-process, backed by an in-memory SQLite,
// and monkeypatches global fetch so the unmodified client talks to it as if it
// were the live api.getstub.dev. This lets the reference run and be verified
// with no network. Point the client at the real registry (its default) when
// you run outside the sandbox.
//
// This imports the ACTUAL registry.js and local-db.js from the backend, so
// what it proves here is what the deployed Worker does.

import { Registry } from '../../backend/registry.js';
import { makeLocalDb } from '../../backend/local-db.js';

export async function startLocalRegistry() {
  const db = makeLocalDb();
  const witness = await crypto.subtle.generateKey({ name: 'Ed25519' }, true, ['sign', 'verify']);
  const reg = new Registry(db, witness.privateKey);

  const realFetch = globalThis.fetch;
  globalThis.fetch = async (url, opts = {}) => {
    const u = String(url);
    const path = u.replace(/^https?:\/\/[^/]+/, '');
    const body = opts.body ? JSON.parse(opts.body) : {};
    const json = (obj, status = 200) => new Response(JSON.stringify(obj), {
      status, headers: { 'Content-Type': 'application/json' },
    });

    if (path === '/register' && opts.method === 'POST') {
      return json(await reg.registerOperator(body));
    }
    if (path === '/issue' && opts.method === 'POST') {
      return json(await reg.issue(body));
    }
    if (path.startsWith('/check/') || path.startsWith('/resolve/')) {
      const id = path.split('/').pop();
      const r = await reg.resolve(id);
      return json(r);
    }
    if (path === '/health' || path === '') {
      return json({ ok: true, service: 'stub-registry-local', version: '0.1' });
    }
    // fall through to the real network for anything else
    return realFetch(url, opts);
  };

  return { reg, restore: () => { globalThis.fetch = realFetch; } };
}
